### PR TITLE
Update contributing guide

### PR DESCRIPTION
**Pull request by dnephin**
[Preview](https://github.com/hashicorp/consul/blob/dnephin/update-contrib/.github/CONTRIBUTING.md) the changes.

* Replace the mailing list link with Discuss. The mailing list has not been used for a few years.
* `make format` no longer exists, so replace it with `gofmt`
* Remove the instructions about running all tests locally, and debugging flakes.
* Update note about vendoring of dependencies.
* Add link to developer docs.

I suspect the only potentially controversial change here is the guidance around running all tests and finding flaky tests. I believe this is a good change for a few reasons:
1. Tests passing locally is not really relevant. The important signal is that tests pass in a predictable environment (CI), so having contributors fight with our test suite that can fail for a number of reasons is unnecessary.
2. Finding and debugging flaky tests is challenging. The goal of this document is to make it easier for first time contributors. Pointing new contributors at this advanced workflow is more likely to be a distraction.
3. We rarely (maybe never) use this makefile target. It is likely broken, and it is not the method we are planning on using for detecting flakes in the future. So the sooner we remove it from the docs, the easier it will be for us to remove it from the Makefile later.

### CODE REVIEW

File: .github/CONTRIBUTING.md
Comment (kisunji) [inline .github/CONTRIBUTING.md:100]: Is this true? I think I've seen some PRs require approval from a `member` to run some CI tests (maybe to prevent malicious test runs)

**Review by kisunji — APPROVED** [review]
Just one potential clarification; otherwise LGTM

File: .github/CONTRIBUTING.md
Comment (dnephin) [reply .github/CONTRIBUTING.md:100]: Good question. Github has added a prompt for first time contributors that requires a member to click it. But I believe that only applies to github actions. I think I've seen CircleCI checks run even without that button being clicked. It's possible I don't fully understand how that new button works, but either way, one way or another we will use CI to verify the change.

There are some cases where CircleCI just doesn't seem to want to run anything (ex: #10701), but I think that is due to a bug. Normally CI should run on all PRs.

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/437919.

**Pull request merged**
